<h4>Available annotation tools</h4>

<div class="card">
    <div class="card-header">
        <h5>IgBLAST</h5>
    </div>
    <div class="card-body">
        <div class="row">
            <div class="col-2">
                <div class="row">
                    <div class="col-12">
                        <strong>Available versions:</strong>
                    </div>
                    <div class="col-12">
                        IgBLAST 14
                    </div>

                    <div class="col-12">
                        IgBLAST 16
                    </div>
                </div>

            </div>
            <div class="col-2">
                <div class="row">
                    <div class="col-12">
                        <strong>Official website:</strong>
                    </div>
                    <div class="col-12">
                        https://www.ncbi.nlm.nih.gov/igblast/
                    </div>

                    <div class="col-12">
                        <strong>Maintainer in this platform:</strong>
                    </div>
                    <div class="col-12">
                        Erand Smakaj <a href="mailto:erand.smakaj@gmail.com">erand.smakaj@fhnw.ch</a>
                    </div>
                </div>
            </div>
            <div class="col-8">
                The variable domain of an immunoglobulin (IG) sequence is encoded by multiple genes, including the variable (V) gene, the diversity (D) gene and the joining (J) gene. Analysis of IG sequences typically requires identification of each gene, as well as a comparison of sequence variations in the context of defined regions. General purpose tools, such as the BLAST program, have only limited use for such tasks, as the rearranged nature of an IG sequence and the variable length of each gene requires multiple rounds of BLAST searches for a single IG sequence. Additionally, manual assembly of different genes is difficult and error-prone. To address these issues and to facilitate other common tasks in analysing IG sequences, we have developed the sequence analysis tool IgBLAST (http://www.ncbi.nlm.nih.gov/igblast/). With this tool, users can view the matches to the germline V, D and J genes, details at rearrangement junctions, the delineation of IG V domain framework regions and complementarity determining regions. IgBLAST has the capability to analyse nucleotide and protein sequences and can process sequences in batches. Furthermore, IgBLAST allows searches against the germline gene databases and other sequence databases simultaneously to minimize the chance of missing possibly the best matching germline V gene.
            </div>
        </div>
    </div>

</div>